File: src/structural/flyweight/python/classes/factories/delivery_factory.py
from classes.delivery_location import DeliveryLocation
from classes.interfaces.delivery_flyweight_protocol import \
    DeliveryFlyweightProtocol
from classes.interfaces.delivery_location_data import DeliveryLocationData


class DeliveryFactory:
    def __init__(self) -> None:
        self.locations: dict[str, DeliveryLocation] = {}

    @staticmethod
    def __create_id(data: DeliveryLocationData) -> str:
        return '_'.join(map(lambda value: value.replace(' ', '')
                            .lower(), [data.city, data.street]))

    def create_location(
        self,
        intrinsicState: DeliveryLocationData
    ) -> DeliveryFlyweightProtocol:

        key = self.__create_id(intrinsicState)
        if (key not in self.locations.keys()):
            self.locations[key] = DeliveryLocation(intrinsicState)
        return self.locations[key]

    def get_locations(self) -> dict[str, DeliveryLocation]:
        return dict(self.locations)

    def __repr__(self) -> str:
        return f'{self.__class__.__name__} {self.__dict__}'

File: src/structural/flyweight/typescript/classes/factories/DeliveryFactory.ts
import { DeliveryLocation } from '../DeliveryLocation';
import { DeliveryFlyweight as DeliveryFlyweightProtocol } from '../interfaces/DeliveryFlyweight';
import { DeliveryLocationData } from '../interfaces/DeliveryLocationData';

export class DeliveryFactory {
  private readonly _locations: Record<string, DeliveryLocation> = {};

  private createId(data: DeliveryLocationData): string {
    return [data.city, data.street]
      .map((value: string) => {
        return value.replace(/\s/g, '').toLowerCase();
      })
      .join('_');
  }

  get locations(): Record<string, DeliveryLocation> {
    return this._locations;
  }

  createLocation(
    intrinsicState: DeliveryLocationData,
  ): DeliveryFlyweightProtocol {
    const key = this.createId(intrinsicState);
    if (!this._locations[key])
      this._locations[key] = new DeliveryLocation(intrinsicState);

    return this._locations[key];
  }
}

const d = new DeliveryFactory();
d.createLocation({ city: 'Joinville', street: 'Inambu' });
